### ISSUE TITLE
Make Rename and Code Actions optionally apply changes to the workspace

### ISSUE THREAD

**Issue by nosami** [OmniSharp/omnisharp-roslyn#225]
When using editor events to update the server with changes, it's much simpler to have the following workflow :-
```
editor -> ask server for changes -> apply changes on editor -> editor events propagate the changes back to the server.
```
When changes are applied automatically, the editor needs to distinguish between user applied changes and changes that need to be applied due to a refactoring so as not to send changes back to the server that have already been applied.

I propose that we add a flag onto the request (or reuse existing WantsChanges flag?) to make applying changes optional.

Thoughts? //cc @jrieken

**Comment by davidfowl** [comment]
What about refactorings that span across the entire project?

**Comment by nosami** [comment]
Same thing applies. The editor will send the changes back once they have been applied.

**Comment by jrieken** [comment]
I like. It's related to #112 and something I have on my list for a long time. As Pilchie already mentioned, we should differentiate between _open_ and _closed_ files. I think _closed_ files can be modified on disk straight away but for _open_ files we should generate changes - either small text changes or the big buffer update. I can sign up for this and propose some code.

**Comment by jrieken** [comment]
more resources: Workspace#TryApplyChanges (http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces) and how MSBuildWorkspace does it (ApplyTextDocumentChanges): http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces.Desktop/Workspace/MSBuild/MSBuildWorkspace.cs,993

**Comment by nosami** [comment]
Don't you need to open the closed files though? You'd want to know what changes had been made. You might want to undo.

**Comment by jrieken** [comment]
Fair point about undo...

**Comment by nosami** [comment]
I think that even with the open/closed arguement, it would still be possible to keep this workflow. The editor can decide whether or not to open a closed document and send the changes regardless.

**Comment by jrieken** [comment]
That is true if OmniSharp sends changes for all files (open & closed). Another question is how those changes should be send to the editors? My favourite is the event emitter but then HTTP-editor integrations aren't part of the game anymore. Thoughts?

**Comment by nosami** [comment]
Currently, the only editors that are using changes are @code and atom (stdio). The others (http) all expect the server to update the workspace and send the entire buffer(s) back.

**Comment by nosami** [comment]
Using events sounds like the best option when there are many changes to send, so we may as well use that when we have the option. 

We can still send the entire buffer (or array of changes) to HTTP clients.
